ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -1019,4 +1031,173 @@ void markStartTime() {
+          + " in put request with account and container headers.", RestServiceErrorCode.InvalidAccount);
+    }
+    ensureAccountNameMatch(targetAccount, restRequest);
+    String containerName = getHeader(restRequest.getArgs(), Headers.TARGET_CONTAINER_NAME, false);
+    Container targetContainer = targetAccount.getContainerByName(containerName);
+    if (targetContainer == null) {
+      frontendMetrics.unrecognizedContainerNameCount.inc();
+      throw new RestServiceException(
+          "Container cannot be found for accountName=" + accountName + " and containerName=" + containerName
+              + " in put request with account and container headers.", RestServiceErrorCode.InvalidContainer);
+    }
+    setTargetAccountAndContainerInRestRequest(restRequest, targetAccount, targetContainer);
+  }
+
+  /**
+   * Obtains the target {@link Account} and {@link Container} id from the blobId string, queries the {@link AccountService}
+   * to get the corresponding {@link Account} and {@link Container}, and injects the target {@link Account} and
+   * {@link Container} into the {@link RestRequest}.
+   * @param blobIdStr The blobId string to get the target {@link Account} and {@link Container} id。
+   * @param restRequest The rest request to insert the target {@link Account} and {@link Container}.
+   * @throws RestServiceException if 1) either {@link Account} or {@link Container} could not be found; or 2)
+   *                              either {@link Account} or {@link Container} were explicitly specified as
+   *                              {@link Account#UNKNOWN_ACCOUNT} or {@link Container#UNKNOWN_CONTAINER}.
+   */
+  private void injectTargetAccountAndContainerFromBlobId(String blobIdStr, RestRequest restRequest)
+      throws RestServiceException {
+    BlobId blobId;
+    try {
+      blobId = new BlobId(blobIdStr, clusterMap);
+    } catch (Exception e) {
+      throw new RestServiceException("Invalid blob id=" + blobIdStr, RestServiceErrorCode.BadRequest);
+    }
+    Account targetAccount = accountService.getAccountById(blobId.getAccountId());
+    if (targetAccount == null) {
+      frontendMetrics.getHeadDeleteUnrecognizedAccountCount.inc();
+      // @todo The check can be removed once HelixAccountService is running with UNKNOWN_ACCOUNT created.
+      if (blobId.getAccountId() != Account.UNKNOWN_ACCOUNT_ID) {
+        throw new RestServiceException(
+            "Account from blobId=" + blobIdStr + "with accountId=" + blobId.getAccountId() + " cannot be recognized",
+            RestServiceErrorCode.InvalidAccount);
+      } else {
+        logger.debug(
+            "Account cannot be found for blobId={} with accountId={}. Setting targetAccount to UNKNOWN_ACCOUNT",
+            blobIdStr, blobId.getAccountId());
+        targetAccount = Account.UNKNOWN_ACCOUNT;
+      }
+    }
+    Container targetContainer = targetAccount.getContainerById(blobId.getContainerId());
```

Comment:
This is not the place to choose public or private container. It should be after getting blob info from server, which has not been implemented yet.